### PR TITLE
[hw, sysrst_ctrl] Add test to improve coverage

### PR DESCRIPTION
**Pull request by crteja**
Add test sequence to cover the following scenarios

- Trigger precondition Debounce to Idle state
  - deassert inputs before debounce time
  - disable precondition before debounce time
- Trigger combo detect Debounce to Idle state
  - deassert inputs before debounce time
  - disable detection before debounce time
- Assert ec_rst_l_i before combo action triggers ec_rst_l_o and a valid combination of input transition is detected
- Key intrrrupt control Debounce to idle state transition
  - disable config before debounce time
- Auto block Debounce to Idle state transition
  - disable auto block before debounce time
- ULP debounce to Idle state transition
  - disable ULP feature before debounce time
  - deassert input signal before debounce time

### CODE REVIEW

**Comment by marnovandermaas** [comment]
This PR is a draft until https://github.com/lowRISC/opentitan/pull/17318 is merged.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+  constraint auto_block_debounce_ctl_c {auto_block_debounce_ctl inside {[15 : 50]};}
+
+  constraint auto_block_out_ctl_c {
+    // At least one input is auto-blocked
+    auto_block_out_ctl & 'h3 > 0;
+  }
+
+  constraint ulp_debounce_time_c {ulp_debounce_ctl inside {[15 : 50]};}
+
+
```
Comment (marnovandermaas) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv:74]: Nit: remove one of these line returns.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+    // Set the duration for combo to pressed
+    csr_wr(ral.com_det_ctl[i], mask & set_duration[i]);
+
+    // Set the trigger action
+    csr_wr(ral.com_out_ctl[i], mask & set_action[i]);
+
+    // It takes 2-3 clock cycles to sync the register values
+    cfg.clk_aon_rst_vif.wait_clks(3);
+  endtask
```
Comment (marnovandermaas) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv:101]: Is this repeated code from the precondition test?

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+
+  task drive_ac(uint16_t cycles);
+    cfg.vif.ac_present = 1;
+    cfg.clk_aon_rst_vif.wait_clks(cycles);
+    cfg.vif.ac_present = 0;
+  endtask
+
+  task drive_pwrb(uint16_t cycles);
+    cfg.vif.pwrb_in = 1;
+    cfg.clk_aon_rst_vif.wait_clks(1);
```
Comment (marnovandermaas) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv:345]: Why is this necessary here, but not in `drive_ac`?

**Review by marnovandermaas — COMMENTED** [review]
Thanks for this PR, at first glance these tests look useful to me. Let's wait for the precondition PR to be merged, this PR to be rebased and then I will run some tests to check whether the behavior is as I expect it.

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv (hunk 3 above)
Comment (crteja) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv:345]: ULP detects HIGH level on `ac_present`, whereas for `pwrb_i` a HIGH->LOW transition is detected. So, in this task, first `pwrb_i` is set to HIGH and the transitioned to LOW and then released

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv (hunk 2 above)
Comment (crteja) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv:101]: Yes, most of the task statements are same except the addition of `mask` input

Comment (marnovandermaas) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv:101]: Is there any clean way we can avoid this code repetition?

Comment (crteja) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv]: Yes, derived the new sequence `sysrst_ctrl_feature_disable_vseq` from `sysrst_ctrl_combo_detect_with_pre_cond_vseq` to reuse the task `config_combo_register`. Updated function in sysrst_ctrl interface for better reuse across sequences.
Updated code in [f57e7fc](https://github.com/lowRISC/opentitan/compare/91f0087b4c09a4a2793dc6380ba14973d08cbfe8..f57e7fc77e214f056994aaa1e207d5d78d66e15d)

**Comment by crteja** [comment]
rebased branch to master

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_combo_detect_with_pre_cond_vseq.sv [hunk 4]
```diff
@@ -198,21 +198,24 @@ class sysrst_ctrl_combo_detect_with_pre_cond_vseq extends sysrst_ctrl_base_vseq;
...
-    foreach (ral.com_det_ctl[i]) csr_wr(ral.com_det_ctl[i], set_duration[i]);
+    csr_wr(ral.com_det_ctl[i], mask & set_duration[i]);
 
     // Set the trigger action
-    foreach (ral.com_out_ctl[i]) csr_wr(ral.com_out_ctl[i], set_action[i]);
+    csr_wr(ral.com_out_ctl[i], mask & set_action[i]);
+  endtask
+
+  task config_register();
+    for( int i = 0; i < 4; i++) config_combo_register(i);
```
Comment (marnovandermaas) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_combo_detect_with_pre_cond_vseq.sv:218]: These changes should at least be in a separate commit. Should they be part of this PR?

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+    csr_wr(ral.key_intr_debounce_ctl, set_key_timer);
+    `uvm_info(`gfn, $sformatf("Write data of key_intr_debounce_ctl register:%0d", set_key_timer),
+              UVM_LOW);
+
+    for (int i = 0; i < 4; i++) begin
+      `uvm_info(`gfn, $sformatf("configuring combo channel %0d", i), UVM_LOW)
+      // Scenario : Trigger precondition Debounce to Idle state via
+      //            deassertion of inputs before debounce timer
+      config_combo_register(i);
+      // Wait for 3 clock cycles to sync the register valuee
```
Comment (marnovandermaas) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv:84]: nit: valuee -> value

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+      csr_wr(ral.com_pre_sel_ctl[i], 0);
+      cfg.clk_aon_rst_vif.wait_clks(10);
+      reset_combo_inputs();
+      cfg.clk_aon_rst_vif.wait_clks(3);
+
+      // Scenario : Trigger combo detect Debounce to Idle state via
+      //            deassertion of input signals before debounce timer
+      // Set combo input to trigger precondition
+      config_combo_register(i);
+      // Wait for 3 clock cycles to sync the register valuee
```
Comment (marnovandermaas) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv:108]: nit: valuee -> value

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+      // Disable combo detect before debounce timer
+      csr_wr(ral.com_sel_ctl[i], 0);
+      cfg.clk_aon_rst_vif.wait_clks(15);
+      // Reset combo inputs
+      reset_combo_inputs();
+      cfg.clk_aon_rst_vif.wait_clks(10);
+
+      // Reset combo registers
+      config_combo_register(i, 16'h0);
+      // Wait for 3 clock cycles to sync the register valuee
```
Comment (marnovandermaas) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv:138]: Nit: valuee -> value

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+    end
+  endtask
+
+  task combo_action_ec_rst_l_i_assert();
+    `uvm_info(`gfn, "ec_rst_l_i assertion check for combo detect", UVM_LOW)
+    for (int i = 0; i < 4; i++) begin
+      `uvm_info(`gfn, $sformatf("configuring combo channel %0d", i), UVM_LOW)
+      // Scenario : Assert ec_rst_l_i just before combo action triggers ec_rst_l_o
+      config_combo_register(i);
+      // Wait for 3 clock cycles to sync the register valuee
```
Comment (marnovandermaas) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv:150]: same as above

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+          cfg.clk_aon_rst_vif.wait_clks(5);
+          monitor_ec_rst_low(set_pulse_width);
+        end
+      join
+      `uvm_info(`gfn, "ec_rst_l_o asserted as expected", UVM_LOW)
+      // Reset combo inputs
+      reset_combo_inputs();
+      // Disable combo detect
+      config_combo_register(i, 0);
+      // Wait for 3 clock cycles to sync the register valuee
```
Comment (marnovandermaas) [inline hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv:174]: valuee

**Review by marnovandermaas — COMMENTED** [review]
Thanks for these changes, I think we are still not hitting the code condition mentioned in this issue:
https://github.com/lowRISC/opentitan/issues/12673
Would you mind having a look at that?

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv (hunk 6 above)
Comment (crteja) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv]: Updated in [5053856](https://github.com/lowRISC/opentitan/compare/b882ff878b32a497b4360db8f3da13a7c5249942..505385645b0533cd62b44d840af09d80ce318ccd)

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv (hunk 7 above)
Comment (crteja) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv]: Updated in [5053856](https://github.com/lowRISC/opentitan/compare/b882ff878b32a497b4360db8f3da13a7c5249942..505385645b0533cd62b44d840af09d80ce318ccd)

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv (hunk 8 above)
Comment (crteja) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv]: Updated in [5053856](https://github.com/lowRISC/opentitan/compare/b882ff878b32a497b4360db8f3da13a7c5249942..505385645b0533cd62b44d840af09d80ce318ccd)

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv (hunk 9 above)
Comment (crteja) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_feature_disable_vseq.sv]: Updated in [5053856](https://github.com/lowRISC/opentitan/compare/b882ff878b32a497b4360db8f3da13a7c5249942..505385645b0533cd62b44d840af09d80ce318ccd)

File: hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_combo_detect_with_pre_cond_vseq.sv (hunk 4 above)
Comment (crteja) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_combo_detect_with_pre_cond_vseq.sv:218]: > Thanks for these changes, I think we are still not hitting the code condition mentioned in this issue: #12673 Would you mind having a look at that?

Updated scenarios to cover Detect to Idle state transition in [5053856](https://github.com/lowRISC/opentitan/compare/b882ff878b32a497b4360db8f3da13a7c5249942..505385645b0533cd62b44d840af09d80ce318ccd)
image

Comment (crteja) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_combo_detect_with_pre_cond_vseq.sv:218]: > These changes should at least be in a separate commit. Should they be part of this PR?

Will move the changes to a new commit

Comment (crteja) [reply hw/ip/sysrst_ctrl/dv/env/seq_lib/sysrst_ctrl_combo_detect_with_pre_cond_vseq.sv:218]: > > These changes should at least be in a separate commit. Should they be part of this PR?
> 
> Will move the changes to a new commit

Moved the changes to a new commit https://github.com/lowRISC/opentitan/commit/419f3db56444fe0d3aea75a398226319b6811f2a

**Comment by cindychip** [comment]
If you do not mind rebasing it again. I think the CI issues are fixed

**Comment by crteja** [comment]
Thanks @marnovandermaas  and @cindychip for the review. Rebased changes to latest master

**Comment by cindychip** [comment]
my bad, sorry I totally missed this PR. LGTM

**Review by cindychip — APPROVED** [review]
